actors/builtin/power/power_actor.go

```diff
@@ -409,19 +409,31 @@ func (a Actor) OnMinerWindowedPoStFailure(rt Runtime, params *OnMinerWindowedPoS
 	var claim *Claim
 	var st State
 	rt.State().Transaction(&st, func() interface{} {
+		faulty, err := st.hasDetectedFault(adt.AsStore(rt), minerAddr)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "Failed to check if miner was faulty already: %s", err)
+		}
+
 		if err := st.putDetectedFault(adt.AsStore(rt), minerAddr); err != nil {
 			rt.Abortf(exitcode.ErrIllegalState, "Failed to put miner fault: %v", err)
 		}
 
 		var found bool
-		var err error
 		claim, found, err = st.getClaim(adt.AsStore(rt), minerAddr)
 		if err != nil {
 			rt.Abortf(exitcode.ErrIllegalState, "Failed to get miner power from claimed power table for surprise PoSt failure: %v", err)
 		}
 		if !found {
 			rt.Abortf(exitcode.ErrIllegalState, "Failed to find miner power in claimed power table for surprise PoSt failure")
 		}
+
+		if claim.Power.GreaterThanEqual(ConsensusMinerMinPower) && !faulty {
+			// Ensure we only deduct this once...
+			st.TotalNetworkPower = big.Sub(st.TotalNetworkPower, claim.Power)
+			if err := st.setClaim(adt.AsStore(rt), minerAddr, claim); err != nil {
+				rt.Abortf(exitcode.ErrIllegalState, "failed to rewrite power claim after slashing miner: %s", err)
+			}
```

Comment:
I don't quite get what's going on with re-setting the claim. The claim object hasn't been modified. I don't think it _should_ be modified: we should hold the claim object as power-less-declared-faults, and use the detected fault set to include/exclude in total power etc.

Reply:
yeah, that was leftover from trying something else